rabbitmq/src/main/java/io/micronaut/rabbitmq/intercept/RabbitMQConsumerAdvice.java

```diff
@@ -128,180 +130,203 @@ public void process(BeanDefinition<?> beanDefinition, ExecutableMethod<?, ?> met
                 throw new MessageListenerException(String.format("Failed to retrieve a channel pool named [%s] to register a listener", connection), e);
             }
 
-            Channel channel = getChannel(channelPool);
+            Integer numberOfConsumers = queueAnn.getRequiredValue("numberOfConsumers", Integer.class);
 
             Integer prefetch = queueAnn.get("prefetch", Integer.class).orElse(null);
-            try {
-                if (prefetch != null) {
-                    channel.basicQos(prefetch);
-                }
-            } catch (IOException e) {
-                throw new MessageListenerException(String.format("Failed to set a prefetch count of [%s] on the channel", prefetch), e);
-            }
-
-            ConsumerState state = new ConsumerState();
-            state.channelPool = channelPool;
-            state.consumerTag = clientTag;
-            consumerChannels.put(channel, state);
-
-            Map<String, Object> arguments = new HashMap<>();
-
-            List<AnnotationValue<RabbitProperty>> propertyAnnotations = method.getAnnotationValuesByType(RabbitProperty.class);
-            Collections.reverse(propertyAnnotations); //set the values in the class first so methods can override
-            propertyAnnotations.forEach((prop) -> {
-                String name = prop.getRequiredValue("name", String.class);
-                String value = prop.getValue(String.class).orElse(null);
-                Class type = prop.get("type", Class.class).orElse(null);
-
-                if (StringUtils.isNotEmpty(name) && StringUtils.isNotEmpty(value)) {
-                    if (type != null && type != Void.class) {
-                        Optional<Object> converted = conversionService.convert(value, type);
-                        if (converted.isPresent()) {
-                            arguments.put(name, converted.get());
-                        } else {
-                            throw new MessageListenerException(String.format("Could not convert the argument [%s] to the required type [%s]", name, type));
-                        }
-                    } else {
-                        arguments.put(name, value);
-                    }
-                }
-            });
 
-            io.micronaut.context.Qualifier<Object> qualifer = beanDefinition
-                    .getAnnotationNameByStereotype("javax.inject.Qualifier")
-                    .map(type -> Qualifiers.byAnnotation(beanDefinition, type))
-                    .orElse(null);
+            Map<String, Object> arguments = retrieveArguments(method);
+            Object bean = getExecutableMethodBean(beanDefinition, method);
```

Comment:
As i did a review of the code i had a concern about thread safety.
By using multiple channels and threads to consume messages from a queue, the `@RabbitListener` bean implementation must be thread-safe.
With the current implementation, with only one thread per consumer bean this is not an issue.

Maybe it is sufficient to mention it in the documentaion of `@RabbitListener`.
Another possibility would be to have a separate `@RabbitListener` bean instance per consumer/thread so that the bean itself does not need to care about it. But then this issue just traverse to other beans used by the rabbit listener.

Do i see ghosts or is this something to consider?

Reply:
This is a very good point. AFAIK, the `@RabbitListener` is a `Singleton`, so this is already a requirement. I do not know whether we can influence the creation of beans from this class, or whether we can remove the characteristic of it being a Singleton.

Reply:
Given the feature is opt in, I don't see that as an issue. If you enable concurrent execution, then one should be aware any manipulation of data on the bean is thread safe. It's worth a note in the documentation though.